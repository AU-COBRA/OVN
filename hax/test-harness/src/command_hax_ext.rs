use lazy_static::{__Deref, lazy_static};
use std::{ffi::OsStr, process::Command};

pub trait CommandHaxExt {
    fn hax<I: IntoIterator<Item = S>, S: AsRef<OsStr>>(args: I) -> Self;
}

/// Computes a list of arguments that setup the number of parallel
/// jobs for dune accordingly to environment variable `DUNEJOBS`.
fn dune_jobs_args() -> Vec<String> {
    if let Ok(jobs) = std::env::var("DUNEJOBS") {
        vec!["-j".into(), jobs]
    } else {
        vec![]
    }
}

impl CommandHaxExt for Command {
    fn hax<I: IntoIterator<Item = S>, S: AsRef<OsStr>>(args: I) -> Command {
        use assert_cmd::cargo::cargo_bin;
        use std::path::PathBuf;
        struct Paths {
            driver: PathBuf,
            engine: PathBuf,
            cargo_hax: PathBuf,
        }
        const CARGO_HAX: &str = "cargo-hax";
        lazy_static! {
            static ref PATHS: Option<Paths> = {
                if let "yes" | "y" | "true" | "1" = std::env::var("CARGO_TESTS_ASSUME_BUILT").unwrap_or("".into()).to_lowercase().as_str() {
                    return None;
                }
                let root = std::env::current_dir().unwrap();
                let root = root.parent().unwrap();
                let engine_dir = root.join("engine");
                // Make sure binaries are built. Note this doesn't
                // include `hax-engine-names-extract`: its build
                // script requires the driver and CLI of `hax` to be
                // available.
                assert!(Command::new("cargo")
                            .args(&["build", "--bins"])
                            .current_dir(&root)
                            .status()
                            .unwrap()
                            .success());
                let cargo_hax = cargo_bin(CARGO_HAX);
                let driver = cargo_bin("driver-hax-frontend-exporter");
                // Now the driver & CLI are installed, call `cargo
                // build` injecting their paths
                assert!(Command::new("cargo")
                        .args(&["build", "--workspace", "--bin", "hax-engine-names-extract"])
                        .env("HAX_CARGO_COMMAND_PATH", &cargo_hax)
                        .env("HAX_RUSTC_DRIVER_BINARY", &driver)
                        .current_dir(&root)
                        .status()
                        .unwrap()
                        .success());
                assert!(Command::new("dune")
                        .args(&["build"])
                        .args(dune_jobs_args())
                        .env("HAX_JSON_SCHEMA_EXPORTER_BINARY", cargo_bin("hax-export-json-schemas"))
                        .env("HAX_ENGINE_NAMES_EXTRACT_BINARY", cargo_bin("hax-engine-names-extract"))
                        .current_dir(engine_dir.clone())
                        .status()
                        .unwrap()
                        .success());
                Some(Paths {
                    driver,
                    cargo_hax,
                    engine: engine_dir.join("_build/install/default/bin/hax-engine"),
                })
            };
        }
        let mut cmd = match PATHS.deref() {
            Some(paths) => {
                let mut cmd = Command::new(paths.cargo_hax.clone());
                cmd.env("HAX_RUSTC_DRIVER_BINARY", paths.driver.clone());
                cmd.env("HAX_ENGINE_BINARY", paths.engine.clone());
                cmd
            }
            None => Command::new(CARGO_HAX),
        };
        cmd.args(args);
        // As documented in
        // https://doc.rust-lang.org/cargo/reference/environment-variables.html#dynamic-library-paths,
        // [cargo run] (and thus also [cargo test]) sets dynamic
        // library paths, which causes some issues with dependencies
        // when compiling without rustup
        for env in ["DYLD_FALLBACK_LIBRARY_PATH", "LD_LIBRARY_PATH"] {
            cmd.env_remove(env);
        }
        cmd
    }
}
