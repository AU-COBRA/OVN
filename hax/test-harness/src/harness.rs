mod command_hax_ext;
use command_hax_ext::*;
use serde_json::{Map, Value};
use std::process::{Command, Stdio};

#[derive(Clone, Debug, serde::Serialize)]
pub enum TestKind {
    Translate { backend: String },
}

impl TestKind {
    fn as_name(&self) -> String {
        (match self {
            TestKind::Translate { backend } => ["into".to_string(), backend.clone()],
        })
        .join("-")
    }
}

#[allow(dead_code)]
fn bool_true() -> bool {
    true
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct TestSnapshot {
    #[serde(default = "bool_true")]
    pub stderr: bool,
    #[serde(default = "bool_true")]
    pub stdout: bool,
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct TestSpec {
    /// is the test optional? (useful for slow tests for instance)
    pub optional: bool,
    /// a broken test a test that should succeed (or fail) but does
    /// not dues to a bug to be fixed (see field [issue_id] below)
    pub broken: bool,
    /// Github issue ID
    pub issue_id: Option<u64>,
    /// Is that a positive or a negative test?
    pub positive: bool,
    pub snapshot: TestSnapshot,
    pub include_flag: Option<String>,
    pub backend_options: Option<Vec<String>>,
}

impl From<Value> for TestSpec {
    /// Parse a JSON value into a TestSpec
    fn from(o: Value) -> Self {
        fn as_opt_bool(v: &Value, def: bool) -> Option<bool> {
            if v.is_null() {
                return Some(def);
            }
            v.as_bool()
        }
        fn as_bool(o: &Value, k: &str, def: bool) -> bool {
            let v = &o[k];
            as_opt_bool(v, def)
                .expect(format!("[{}] was expected to be a boolean, got {}", k, v).as_str())
        }
        let snapshot = &o["snapshot"];
        TestSpec {
            optional: as_bool(&o, "optional", false),
            broken: as_bool(&o, "broken", false),
            positive: as_bool(&o, "positive", true),
            issue_id: o["positive"].as_u64(),
            include_flag: o["include-flag"].as_str().map(|s| s.into()),
            backend_options: serde_json::from_value(o["backend-options"].clone()).unwrap(),
            snapshot: as_opt_bool(snapshot, true)
                .map(|b| TestSnapshot {
                    stderr: b,
                    stdout: b,
                })
                .or_else(|| match snapshot.as_str() {
                    Some(v @ ("stdout" | "stderr" | "both" | "none")) => Some(TestSnapshot {
                        stdout: matches!(v, "stdout" | "both"),
                        stderr: matches!(v, "stderr" | "both"),
                    }),
                    Some(v) => panic!(
                        "[snapshot] is \"{}\" but was expected to be \"stderr\", \"stdout\" or \"both\"", v
                    ),
                    None => None,
                })
                .unwrap_or_else(|| TestSnapshot {
                    stderr: as_bool(&snapshot, "stderr", true),
                    stdout: as_bool(&snapshot, "stdout", true),
                }),
        }
    }
}

/// The information for a test is given by `cargo metadata`
#[derive(Clone, Debug, serde::Serialize)]
pub struct TestInfo {
    pub name: String,
    pub manifest: std::path::PathBuf,
    pub description: Option<String>,
}

#[derive(Clone, Debug, serde::Serialize)]
pub struct Test {
    pub kind: TestKind,
    pub info: TestInfo,
    pub spec: TestSpec,
}

impl Test {
    fn as_args(&self) -> Vec<String> {
        match &self.kind {
            TestKind::Translate { backend } => {
                let mut args = vec![];
                args.push("into".to_string());
                if let Some(i) = self.spec.include_flag.as_ref() {
                    args.push("-i".to_string());
                    args.push(i.to_string());
                }
                args.push("--dry-run".to_string());
                args.push(backend.clone());
                if let Some(backend_options) = &self.spec.backend_options {
                    args.extend_from_slice(backend_options.clone().as_slice());
                }
                args
            }
        }
    }
}

impl std::fmt::Display for Test {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "{} - {:?}", self.info.name, self.kind)?;
        if let Some(issue_id) = self.spec.issue_id {
            write!(f, " #{}", issue_id)?;
        };
        Ok(())
    }
}

impl Test {
    fn into_runner(self, workspace: String) -> Result<(), libtest_mimic::Failed> {
        // 1. cook a command
        let mut cmd = Command::hax(&["-C"]);
        cmd.arg("--manifest-path").arg(self.info.manifest.clone());
        cmd.arg(";");
        cmd.stdout(Stdio::piped()).stderr(Stdio::piped());
        cmd.args(self.as_args());

        // 2. execute it (twice, idea of @franziskuskiefer, so that
        // the messages related to building dependencies are not
        // included in the second one)
        let _ = cmd.output().unwrap();
        let out = cmd.output().unwrap();

        let command_successful = out.status.success();
        let cleanup = |s: String| {
            use lazy_static::lazy_static;
            use regex::Regex;
            lazy_static! {
                // Regex [TIME] matches compile times
                static ref TIME: Regex = Regex::new(r"\bin \d+(\.\d+)?s\b").unwrap();
                static ref LOCK: Regex = Regex::new(r"Blocking waiting for \w+ lock on (the registry index|build directory|package cache)").unwrap();
            }
            TIME.replace_all(
                LOCK.replace_all(
                    &s.replace(r"\", "/").replace(&workspace, "WORKSPACE_ROOT"),
                    "",
                )
                .as_ref(),
                "in XXs",
            )
            .trim()
            .to_string()
        };
        let serr = cleanup(String::from_utf8_lossy(&out.stderr).to_string());
        let sout = String::from_utf8_lossy(&out.stdout).to_string();

        // 3. make sure the test is successful
        let mut snapshot: Map<String, Value> = Map::new();
        if self.spec.snapshot.stderr {
            snapshot.insert("stderr".to_string(), Value::String(serr.clone()));
        }
        if self.spec.snapshot.stdout {
            snapshot.insert(
                "stdout".to_string(),
                serde_json::from_str(&sout)
                    .map(|out: hax_types::engine_api::Output| {
                        use serde_json::json;
                        json!({
                            "diagnostics": Value::Array(out.diagnostics.into_iter().map(|diag| json!({
                                "spans": Value::Array(diag.span.clone().into_iter().map(|span| Value::String(format!("{:?}", span))).collect()),
                                "message": Value::String(format!("{}", diag)),
                            })).collect()),
                            "files": Value::Object(out.files.into_iter().map(|file| (file.path, Value::String(file.contents))).collect())
                        })
                    })
                    .unwrap_or_else(|_| Value::String(cleanup(sout.clone()))),
            );
        }

        if !snapshot.is_empty() {
            let exit = out.status.code().unwrap_or(std::i32::MAX);
            snapshot.insert("exit".to_string(), exit.into());
            let snapshot = Value::Object(snapshot);
            let name = format!("{} {}", self.info.name, self.kind.as_name());

            let mut info = self.clone();
            info.info.manifest = info.info.manifest.strip_prefix(workspace).unwrap().into();
            insta::with_settings!({
                info => &info,
            }, { insta::assert_toml_snapshot!(name, snapshot) })
        }

        let err = |s: &str| {
            Err(format!(
                "Command {s}.\nThe command was: {:?}{}",
                cmd,
                if command_successful {
                    "".to_string()
                } else {
                    format!("\nSTDOUT:\n{}\nSTDERR:\n{}", sout, serr)
                }
            ))
        };
        match (command_successful, (self.spec.positive, self.spec.broken)) {
            (true, (true, false) | (false, true)) => Ok(()),
            (false, (false, false) | (true, true)) => Ok(()),
            (false, (false, true)) => err("failed, but this is a negative test marked broken")?,
            (false, (true, false)) => err("failed")?,
            (true, (true, true)) => err("succeeded, but this is a positive test marked broken")?,
            (true, (false, false)) => err("succeeded, but this is a negative test")?,
        }
    }

    fn into_trial(&self, workspace: &String) -> libtest_mimic::Trial {
        libtest_mimic::Trial::test(format!("{}", &self), {
            let test = self.clone();
            let workspace = workspace.clone();
            move || test.clone().into_runner(workspace)
        })
        .with_kind(if self.spec.positive {
            "positive"
        } else {
            "negative"
        })
        .with_ignored_flag(self.spec.optional)
    }
}

/// Given [metadata] the table declared in a test's [Cargo.toml]
/// [workspace.hax-tests], this function returns a list of tests
fn parse_hax_tests_metadata(info: TestInfo, metadata: &Value) -> Vec<Test> {
    if metadata.is_null() {
        return vec![];
    }

    metadata
        .as_object()
        .expect(
            format!(
                "Expected value at key [hax-tests] to be a dictionary for package {:#?}",
                info
            )
            .as_str(),
        )
        .into_iter()
        .flat_map(|(a, o)| {
            o.as_object()
                .expect(
                    format!(
                        "Expected value at key [{}] be a dictionary for package {:#?}",
                        a, info
                    )
                    .as_str(),
                )
                .into_iter()
                .flat_map(|(key, o)| key.split("+").map(|k| (k.trim().to_string(), o.clone())))
                .map(|(b, o)| (a.clone(), b, o))
        })
        .map(|(a, b, o)| Test {
            spec: o.into(),
            info: info.clone(),
            kind: match a.as_str() {
                "into" => TestKind::Translate { backend: b },
                _ => panic!(
                    "unexpected metadata [hax-tests.{}.{}] for package {:#?}",
                    a, b, info
                ),
            },
        })
        .collect()
}

fn main() {
    let metadata = cargo_metadata::MetadataCommand::new()
        .manifest_path("../tests/Cargo.toml")
        .exec()
        .unwrap();
    let workspace_root: String = metadata.workspace_root.into();

    let mut args = libtest_mimic::Arguments::from_args();
    args.test_threads = Some(1);

    libtest_mimic::run(
        &args,
        metadata
            .packages
            .into_iter()
            .flat_map(|o| {
                parse_hax_tests_metadata(
                    TestInfo {
                        name: o.name,
                        description: o.description,
                        manifest: o.manifest_path.into(),
                    },
                    &o.metadata["hax-tests"],
                )
            })
            .map(|test| test.into_trial(&workspace_root))
            .collect(),
    )
    .exit();
}
