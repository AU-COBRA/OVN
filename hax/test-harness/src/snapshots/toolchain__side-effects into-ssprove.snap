---
source: test-harness/src/harness.rs
expression: snapshot
info:
  kind:
    Translate:
      backend: ssprove
  info:
    name: side-effects
    manifest: side-effects/Cargo.toml
    description: ~
  spec:
    optional: false
    broken: false
    issue_id: ~
    positive: true
    snapshot:
      stderr: true
      stdout: true
    include_flag: ~
    backend_options: ~
---
exit = 0
stderr = 'Finished `dev` profile [unoptimized + debuginfo] target(s) in XXs'

[stdout]
diagnostics = []

[stdout.files]
"Side_effects.v" = '''
(* File automatically generated by Hacspec *)
Set Warnings "-notation-overridden,-ambiguous-paths".
From Crypt Require Import choice_type Package Prelude.
Import PackageNotation.
From extructures Require Import ord fset.
From mathcomp Require Import word_ssrZ word.
From Jasmin Require Import word.

From Coq Require Import ZArith.
From Coq Require Import Strings.String.
Import List.ListNotations.
Open Scope list_scope.
Open Scope Z_scope.
Open Scope bool_scope.

From Hacspec Require Import ChoiceEquality.
From Hacspec Require Import LocationUtility.
From Hacspec Require Import Hacspec_Lib_Comparable.
From Hacspec Require Import Hacspec_Lib_Pre.
From Hacspec Require Import Hacspec_Lib.

Open Scope hacspec_scope.
Import choice.Choice.Exports.

Obligation Tactic := (* try timeout 8 *) solve_ssprove_obligations.

(*Not implemented yet? todo(item)*)

Equations add3 {L1 : {fset Location}} {L2 : {fset Location}} {L3 : {fset Location}} {I1 : Interface} {I2 : Interface} {I3 : Interface} (x : both L1 I1 int32) (y : both L2 I2 int32) (z : both L3 I3 int32) : both (L1 :|: L2 :|: L3) (I1 :|: I2 :|: I3) int32 :=
  add3 x y z  :=
    solve_lift (impl__u32__wrapping_add (impl__u32__wrapping_add x y) z) : both (L1 :|: L2 :|: L3) (I1 :|: I2 :|: I3) int32.
Fail Next Obligation.

Equations direct_result_question_mark {L1 : {fset Location}} {I1 : Interface} (y : both L1 I1 (t_Result 'unit int32)) : both L1 I1 (t_Result int8 int32) :=
  direct_result_question_mark y  :=
    solve_lift (run (letm[choice_typeMonad.result_bind_code int32] _ := y in
    Result_Ok (Result_Ok (ret_both (0 : int8))))) : both L1 I1 (t_Result int8 int32).
Fail Next Obligation.

Equations direct_result_question_mark_coercion {L1 : {fset Location}} {I1 : Interface} (y : both L1 I1 (t_Result int8 int16)) : both L1 I1 (t_Result int8 int32) :=
  direct_result_question_mark_coercion y  :=
    solve_lift (run (letm[choice_typeMonad.result_bind_code int32] hoist1 := impl__map_err y f_from in
    Result_Ok (Result_Ok hoist1))) : both L1 I1 (t_Result int8 int32).
Fail Next Obligation.

Equations early_returns {L1 : {fset Location}} {I1 : Interface} (x : both L1 I1 int32) : both L1 I1 int32 :=
  early_returns x  :=
    solve_lift (run (letm[choice_typeMonad.result_bind_code int32] _ := ifb x >.? (ret_both (3 : int32))
    then letm[choice_typeMonad.result_bind_code int32] hoist2 := ControlFlow_Break (ret_both (0 : int32)) in
    ControlFlow_Continue (never_to_any hoist2)
    else () in
    letb hoist3 := x >.? (ret_both (30 : int32)) in
    letm[choice_typeMonad.result_bind_code int32] hoist5 := ifb hoist3
    then matchb ret_both (true : 'bool) with
    | true =>
      letm[choice_typeMonad.result_bind_code int32] hoist4 := ControlFlow_Break (ret_both (34 : int32)) in
      ControlFlow_Continue (solve_lift (never_to_any hoist4))
    | _ =>
      ControlFlow_Continue (solve_lift (ret_both (3 : int32)))
    end
    else ControlFlow_Continue (letb _ := assign todo(term) in
    x .+ (ret_both (1 : int32))) in
    letb hoist6 := impl__u32__wrapping_add (ret_both (123 : int32)) hoist5 in
    letb hoist7 := impl__u32__wrapping_add hoist6 x in
    letm[choice_typeMonad.result_bind_code int32] hoist8 := ControlFlow_Break hoist7 in
    ControlFlow_Continue (never_to_any hoist8))) : both L1 I1 int32.
Fail Next Obligation.

Definition y_loc : Location :=
  (int32;0%nat).
Definition y_loc : Location :=
  (int32;1%nat).
Equations local_mutation {L1 : {fset Location}} {I1 : Interface} (x : both L1 I1 int32) : both (L1 :|: fset [y_loc;y_loc]) I1 int32 :=
  local_mutation x  :=
    letb y loc(y_loc) := ret_both (0 : int32) in
    letb _ := assign todo(term) in
    letb hoist9 := x >.? (ret_both (3 : int32)) in
    solve_lift (ifb hoist9
    then letb _ := assign todo(term) in
    letb y loc(y_loc) := x ./ (ret_both (2 : int32)) in
    letb _ := assign todo(term) in
    letb hoist10 := ret_both (0 : int32) in
    letb hoist11 := Build_t_Range (f_start := hoist10) (f_end := ret_both (10 : int32)) in
    letb hoist12 := f_into_iter hoist11 in
    letb _ := foldi_both_list hoist12 (fun i =>
      ssp (fun _ =>
        assign todo(term) : (both (*0*)(L1:|:fset []) (I1) 'unit))) (ret_both (tt : 'unit)) in
    impl__u32__wrapping_add x y
    else letb hoist15 := matchb x with
    | 12 =>
      letb _ := assign todo(term) in
      solve_lift (ret_both (3 : int32))
    | 13 =>
      letb hoist14 := x in
      letb _ := assign todo(term) in
      letb hoist13 := impl__u32__wrapping_add (ret_both (123 : int32)) x in
      solve_lift (add3 hoist14 hoist13 x)
    | _ =>
      solve_lift (ret_both (0 : int32))
    end in
    letb _ := assign todo(term) in
    impl__u32__wrapping_add x y) : both (L1 :|: fset [y_loc;y_loc]) I1 int32.
Fail Next Obligation.

Equations options {L1 : {fset Location}} {L2 : {fset Location}} {L3 : {fset Location}} {I1 : Interface} {I2 : Interface} {I3 : Interface} (x : both L1 I1 (t_Option int8)) (y : both L2 I2 (t_Option int8)) (z : both L3 I3 (t_Option int64)) : both (L1 :|: L2 :|: L3) (I1 :|: I2 :|: I3) (t_Option int8) :=
  options x y z  :=
    solve_lift (run (letm[choice_typeMonad.option_bind_code] hoist19 := x in
    letb hoist20 := hoist19 >.? (ret_both (10 : int8)) in
    letm[choice_typeMonad.option_bind_code] hoist26 := ifb hoist20
    then letm[choice_typeMonad.option_bind_code] hoist21 := x in
    Option_Some (letb hoist22 := impl__u8__wrapping_add hoist21 (ret_both (3 : int8)) in
    Option_Some hoist22)
    else letm[choice_typeMonad.option_bind_code] hoist24 := x in
    letm[choice_typeMonad.option_bind_code] hoist23 := y in
    Option_Some (letb hoist25 := impl__u8__wrapping_add hoist24 hoist23 in
    Option_Some hoist25) in
    letm[choice_typeMonad.option_bind_code] hoist27 := hoist26 in
    letm[choice_typeMonad.option_bind_code] v := matchb hoist27 with
    | 3 =>
      Option_None
    | 4 =>
      letm[choice_typeMonad.option_bind_code] hoist16 := z in
      Option_Some (letb hoist17 := hoist16 >.? (ret_both (4 : int64)) in
      letb hoist18 := ifb hoist17
      then ret_both (0 : int8)
      else ret_both (3 : int8) in
      solve_lift ((ret_both (4 : int8)) .+ hoist18))
    | _ =>
      Option_Some (solve_lift (ret_both (12 : int8)))
    end in
    letm[choice_typeMonad.option_bind_code] hoist28 := x in
    letb hoist30 := impl__u8__wrapping_add v hoist28 in
    letm[choice_typeMonad.option_bind_code] hoist29 := y in
    Option_Some (letb hoist31 := impl__u8__wrapping_add hoist30 hoist29 in
    Option_Some hoist31))) : both (L1 :|: L2 :|: L3) (I1 :|: I2 :|: I3) (t_Option int8).
Fail Next Obligation.

Definition y_loc : Location :=
  (int32;2%nat).
Equations question_mark {L1 : {fset Location}} {I1 : Interface} (x : both L1 I1 int32) : both (L1 :|: fset [y_loc]) I1 (t_Result int32 int32) :=
  question_mark x  :=
    solve_lift (run (letm[choice_typeMonad.result_bind_code int32] _ := ifb x >.? (ret_both (40 : int32))
    then letb y loc(y_loc) := ret_both (0 : int32) in
    letb _ := assign todo(term) in
    letb _ := assign todo(term) in
    letb _ := assign todo(term) in
    letb hoist32 := x >.? (ret_both (90 : int32)) in
    ifb hoist32
    then impl__map_err (Result_Err (ret_both (12 : int8))) f_from
    else ()
    else () in
    Result_Ok (Result_Ok (impl__u32__wrapping_add (ret_both (3 : int32)) x)))) : both (L1 :|: fset [y_loc]) I1 (t_Result int32 int32).
Fail Next Obligation.

Definition t_A : choice_type :=
  'unit.
Equations Build_t_A : both (fset []) (fset []) (t_A) :=
  Build_t_A  :=
    solve_lift (ret_both (tt (* Empty tuple *) : (t_A))) : both (fset []) (fset []) (t_A).
Fail Next Obligation.

Definition t_B : choice_type :=
  'unit.
Equations Build_t_B : both (fset []) (fset []) (t_B) :=
  Build_t_B  :=
    solve_lift (ret_both (tt (* Empty tuple *) : (t_B))) : both (fset []) (fset []) (t_B).
Fail Next Obligation.

Definition t_Bar : choice_type :=
  ('bool × nseq ('bool × 'bool) 6 × 'bool).
Equations f_a {L : {fset Location}} {I : Interface} (s : both L I t_Bar) : both L I 'bool :=
  f_a s  :=
    bind_both s (fun x =>
      solve_lift (ret_both (fst x : 'bool))) : both L I 'bool.
Fail Next Obligation.
Equations f_b {L : {fset Location}} {I : Interface} (s : both L I t_Bar) : both L I (nseq ('bool × 'bool) 6 × 'bool) :=
  f_b s  :=
    bind_both s (fun x =>
      solve_lift (ret_both (snd x : (nseq ('bool × 'bool) 6 × 'bool)))) : both L I (nseq ('bool × 'bool) 6 × 'bool).
Fail Next Obligation.
Equations Build_t_Bar {L0 : {fset Location}} {L1 : {fset Location}} {I0 : Interface} {I1 : Interface} {f_a : both L0 I0 'bool} {f_b : both L1 I1 (nseq ('bool × 'bool) 6 × 'bool)} : both (L0:|:L1) (I0:|:I1) (t_Bar) :=
  Build_t_Bar  :=
    bind_both f_b (fun f_b =>
      bind_both f_a (fun f_a =>
        solve_lift (ret_both ((f_a,f_b) : (t_Bar))))) : both (L0:|:L1) (I0:|:I1) (t_Bar).
Fail Next Obligation.
Notation "'Build_t_Bar' '[' x ']' '(' 'f_a' ':=' y ')'" := (Build_t_Bar (f_a := y) (f_b := f_b x)).
Notation "'Build_t_Bar' '[' x ']' '(' 'f_b' ':=' y ')'" := (Build_t_Bar (f_a := f_a x) (f_b := y)).

Equations monad_lifting {L1 : {fset Location}} {I1 : Interface} (x : both L1 I1 int8) : both L1 I1 (t_Result t_A t_B) :=
  monad_lifting x  :=
    solve_lift (run (ifb x >.? (ret_both (123 : int8))
    then letm[choice_typeMonad.result_bind_code (t_Result t_A t_B)] hoist33 := ControlFlow_Continue (Result_Err B) in
    letb hoist34 := Result_Ok hoist33 in
    letm[choice_typeMonad.result_bind_code (t_Result t_A t_B)] hoist35 := ControlFlow_Break hoist34 in
    ControlFlow_Continue (never_to_any hoist35)
    else ControlFlow_Continue (Result_Ok A))) : both L1 I1 (t_Result t_A t_B).
Fail Next Obligation.

Definition t_Foo : choice_type :=
  ('bool × 'bool × t_Vec t_Bar t_Global × nseq t_Bar 6 × t_Bar).
Equations f_x {L : {fset Location}} {I : Interface} (s : both L I t_Foo) : both L I 'bool :=
  f_x s  :=
    bind_both s (fun x =>
      solve_lift (ret_both (fst (fst (fst x)) : 'bool))) : both L I 'bool.
Fail Next Obligation.
Equations f_y {L : {fset Location}} {I : Interface} (s : both L I t_Foo) : both L I ('bool × t_Vec t_Bar t_Global) :=
  f_y s  :=
    bind_both s (fun x =>
      solve_lift (ret_both (snd (fst (fst x)) : ('bool × t_Vec t_Bar t_Global)))) : both L I ('bool × t_Vec t_Bar t_Global).
Fail Next Obligation.
Equations f_z {L : {fset Location}} {I : Interface} (s : both L I t_Foo) : both L I (nseq t_Bar 6) :=
  f_z s  :=
    bind_both s (fun x =>
      solve_lift (ret_both (snd (fst x) : (nseq t_Bar 6)))) : both L I (nseq t_Bar 6).
Fail Next Obligation.
Equations f_bar {L : {fset Location}} {I : Interface} (s : both L I t_Foo) : both L I t_Bar :=
  f_bar s  :=
    bind_both s (fun x =>
      solve_lift (ret_both (snd x : t_Bar))) : both L I t_Bar.
Fail Next Obligation.
Equations Build_t_Foo {L0 : {fset Location}} {L1 : {fset Location}} {L2 : {fset Location}} {L3 : {fset Location}} {I0 : Interface} {I1 : Interface} {I2 : Interface} {I3 : Interface} {f_x : both L0 I0 'bool} {f_y : both L1 I1 ('bool × t_Vec t_Bar t_Global)} {f_z : both L2 I2 (nseq t_Bar 6)} {f_bar : both L3 I3 t_Bar} : both (L0:|:L1:|:L2:|:L3) (I0:|:I1:|:I2:|:I3) (t_Foo) :=
  Build_t_Foo  :=
    bind_both f_bar (fun f_bar =>
      bind_both f_z (fun f_z =>
        bind_both f_y (fun f_y =>
          bind_both f_x (fun f_x =>
            solve_lift (ret_both ((f_x,f_y,f_z,f_bar) : (t_Foo))))))) : both (L0:|:L1:|:L2:|:L3) (I0:|:I1:|:I2:|:I3) (t_Foo).
Fail Next Obligation.
Notation "'Build_t_Foo' '[' x ']' '(' 'f_x' ':=' y ')'" := (Build_t_Foo (f_x := y) (f_y := f_y x) (f_z := f_z x) (f_bar := f_bar x)).
Notation "'Build_t_Foo' '[' x ']' '(' 'f_y' ':=' y ')'" := (Build_t_Foo (f_x := f_x x) (f_y := y) (f_z := f_z x) (f_bar := f_bar x)).
Notation "'Build_t_Foo' '[' x ']' '(' 'f_z' ':=' y ')'" := (Build_t_Foo (f_x := f_x x) (f_y := f_y x) (f_z := y) (f_bar := f_bar x)).
Notation "'Build_t_Foo' '[' x ']' '(' 'f_bar' ':=' y ')'" := (Build_t_Foo (f_x := f_x x) (f_y := f_y x) (f_z := f_z x) (f_bar := y)).

Equations assign_non_trivial_lhs {L1 : {fset Location}} {I1 : Interface} (foo : both L1 I1 t_Foo) : both L1 I1 t_Foo :=
  assign_non_trivial_lhs foo  :=
    letb _ := assign todo(term) in
    letb _ := assign todo(term) in
    letb _ := assign todo(term) in
    letb _ := assign todo(term) in
    letb _ := assign todo(term) in
    solve_lift foo : both L1 I1 t_Foo.
Fail Next Obligation.
'''
